src/dmd/filecache.d

```diff
@@ -0,0 +1,137 @@
+ * Source:      $(LINK2 https://github.com/dlang/dmd/blob/master/src/dmd/filecache.d, _errors.d)
+ * Documentation:  https://dlang.org/phobos/dmd_filecache.html
+ * Coverage:    https://codecov.io/gh/dlang/dmd/src/master/src/dmd/filecache.d
+ */
+
+module dmd.filecache;
+
+import dmd.root.stringtable;
+import dmd.root.array;
+import dmd.root.file;
+
+import core.stdc.stdio;
+
+// A simple wrapper around a string in D for extern(C++) compatibility
+extern(C++) struct Dstring
+{
+    char* ptr;
+    size_t length;
+    extern(D) string toString()
+    {
+        return cast(string) ptr[0 .. length];
+    }
+}
+
+/**
+A line-by-line representation of a $(REF File, dmd, root, file).
+*/
+class FileAndLines
+{
+    File* file;
+    Array!Dstring* lines;
+
+    /**
+    File to read and split into its lines.
+    */
+    this(const(char)* filename)
+    {
+        file = File.create(filename);
+        lines = new Array!Dstring;
+        readAndSplit();
+    }
+
+    // Read a file and split the file buffer linewise
+    private void readAndSplit()
+    {
+        file.read();
+        auto buf = file.buffer;
+        // slice into lines
+        while (*buf)
+        {
+            auto prevBuf = buf;
+            for (; *buf != '\n'; buf++)
+            {
+                if (!*buf)
+                    break;
+            }
+            if (buf != prevBuf)
+                lines.push(Dstring(cast(char*) prevBuf, buf - prevBuf));
+            buf++;
+        }
+    }
+
+    void destroy()
+    {
+        if (file)
+        {
+            file.destroy();
+            file = null;
+            lines.destroy();
+            lines = null;
+        }
+    }
+    ~this()
+    {
+        destroy();
+    }
+}
+
+/**
+A simple file cache that can be used to avoid reading the same file multiple times.
+It stores its cached files as $(LREF FileAndLines)
+*/
+struct FileCache
+{
+    private StringTable files;
+
+    /**
+    Add or get a file from the file cache.
+    If the file isn't part of the cache, it will be read from the filesystem.
+    If the file has been read before, the cached file object will be returned
+
+    Params:
+        file = file to load in (or get from) the cache
+
+    Returns: a $(LREF FileAndLines) object containing a line-by-line representation of the requested file
+    */
+    FileAndLines addOrGetFile(const(char)[] file)
+    {
+        if (auto payload = files.lookup(file.ptr, file.length))
+            if (payload !is null)
+                return cast(typeof(return)) payload.ptrvalue;
+
+        auto lines = new FileAndLines(file.ptr);
+        auto p = files.insert(file.ptr, file.length, cast(void*) lines);
+        return cast(typeof(return)) p.ptrvalue;
+    }
+
+    void initialize()
+    {
+        files._init();
+    }
+
+    void deinitialize()
+    {
+        foreach (sv; files)
+            sv.destroy();
+        files.reset();
+    }
+}
+
+static fileCache = FileCache();
+
+shared static this()
```

Comment:
I think we should avoid module constructors.

Reply:
Removed the module constructor, but now we need to manually initialize in `mars.d` / `frontend.d`